assets/js/src/settings/store/normalize_settings.ts

```diff
@@ -0,0 +1,219 @@
+      'manual',
+      '1and1',
+      'bluehost',
+      'df',
+      'dreamhost',
+      'free',
+      'froghost',
+      'godaddy',
+      'goneo',
+      'googleapps',
+      'greengeeks',
+      'hawkhost',
+      'hivetec',
+      'hostgator',
+      'hosting2go',
+      'hostmonster',
+      'infomaniak',
+      'justhost',
+      'laughingsquid',
+      'lunarpages',
+      'mediatemple',
+      'netfirms',
+      'netissime',
+      'one',
+      'ovh',
+      'phpnet',
+      'planethoster',
+      'rochen',
+      'site5',
+      'siteground',
+      'synthesis',
+      'techark',
+      'vexxhost',
+      'vps',
+      'webcity',
+      'westhost',
+      'wpwebhost',
+    ], 'manual'),
+    mailpoet_sending_frequency: asEnum(['auto', 'manual'], 'manual'),
+    signup_confirmation: asObject({
+      enabled: enabledRadio,
+      subject: text,
+      body: text,
+    }),
+    woocommerce: asObject({
+      transactional_email_id: text,
+      optin_on_checkout: asObject({
+        enabled: enabledRadio,
+        message: text,
+      }),
+      accept_cookie_revenue_tracking: asObject({
+        enabled: disabledRadio,
+        set: enabledRadio,
+      }),
+    }),
+    mailpoet_subscribe_old_woocommerce_customers: asObject({
+      enabled: enabledRadio,
+    }),
+    premium: asObject({
+      premium_key: text,
+      premium_key_state: asObject({
+        state: asEnum([
+          'valid',
+          'invalid',
+          'expiring',
+          'already_used',
+          'check_error',
+        ], 'check_error'),
+        data: asIs,
+      }),
+    }),
+  });
+  return settingsSchema(data) as Settings;
+}
+
+const asString = (defaultValue: string) => (value: any): string => {
+  if (value === undefined) return defaultValue;
+  if (!value) return '';
+  return `${value}`;
+};
+
+const asStringArray = (defaultValue: string[]) => (value: any): string[] => {
+  if (!_.isArray(value)) return defaultValue;
+  return value.map(asString(''));
+};
+
+function asBoolean<T, F>(trueValue: T, falseValue: F, defaultValue: T | F) {
+  return (value: any): T | F => {
+    if (value === undefined) return defaultValue;
+    if (value) return trueValue;
+    return falseValue;
+  };
+}
+
+function asEnum(choices: string[], defaultValue: string) {
+  return (value: any): string => {
+    if (!choices.includes(value)) return defaultValue;
+    return value;
+  };
+}
+
+type Schema = {
+  [key: string]: ReturnType<
+    | typeof asString
+    | typeof asStringArray
+    | typeof asBoolean
+    | typeof asEnum
+    | typeof asObject
+  >
+}
+type SchemaResult<T extends Schema> = {
+  [key in keyof T]: ReturnType<T[key]>
+}
+function asObject<T extends Schema>(schema: T) {
+  return (value: any): SchemaResult<T> => {
+    const object = Object.keys(schema).reduce((result, field) => ({
+      [field]: schema[field](value ? value[field] : undefined),
+      ...result,
+    }), {});
+    return object as SchemaResult<T>;
+  };
+}
```

Comment:
Wow, this is elegant. A bit hard to grasp with the generics and return types but nice!

Reply:
The power of Typescript :D